doc/serdi.1
.TH SERDI 1 "05 May 2019"

.SH NAME
.B serdi \- Read and write RDF syntax

.SH SYNOPSIS
serdi [\fIOPTION\fR]... \fIINPUT\fR...

.SH OPTIONS

.TP
.BR \-I
Set input base URI.

.TP
.BR \-V
Validate inputs.
All necessary data, including schemas, must be passed as inputs.
Output will only be written if validation succeeds.
Combine with -o empty to suppress output and only show validation errors.

.TP
.BR \-a
Write ASCII output if possible.

.TP
.BR \-b
Fast bulk output for large serialisations.

.TP
.BR \-c " " \fIPREFIX\fR
Chop \fIPREFIX\fR from matching blank node IDs.

.TP
.BR \-e
Eat input one character at a time, rather than a page at a time which is the default.
This is useful when reading from a pipe since output will be generated immediately as input arrives, rather than waiting until an entire page of input has arrived.
With this option serdi uses one page less memory, but will likely be significantly slower.

.TP
.BR \-f
Keep full URIs in input (don't qualify with namespace prefixes or make URIs relative).

.TP
.BR \-h
Print the command line options.

.TP
.BR \-i " " \fISYNTAX\fR
Read input as \fISYNTAX\fR.
Valid values (case-insensitive): \*(lqturtle\*(rq, \*(lqntriples\*(rq, \*(lqtrig\*(rq, \*(lqnquads\*(rq.

.TP
.BR \-k " " \fIBYTES\fR
Parser stack size.

.TP
.BR \-l
Lax (non-strict) parsing.

.TP
.BR \-m
Build and serialise a model.
This loads the complete input data into memory before writing the output.
This will normalize statement order, at the cost of performance and memory consumption.

.TP
.BR \-o " " \fISYNTAX\fR
Write output as \fISYNTAX\fR.
Valid values (case-insensitive): \*(lqturtle\*(rq, \*(lqntriples\*(rq, \*(lqtrig\*(rq, \*(lqnquads\*(rq, \*(lqempty\*(rq.
When \*(lqempty\*(rq is given, output is suppressed, so only errors will be printed.

.TP
.BR \-p " " \fIPREFIX\fR
Add \fIPREFIX\fR to blank node IDs.

.TP
.BR \-q
Suppress all output except data.

.TP
.BR \-r " " \fIROOT_URI\fR
Keep relative URIs within \fIROOT_URI\fR.

.TP
.BR \-S
Stream model quickly without inlining.
This only has an effect when a model is used, and disables searching and statement reordering for pretty printing.
Statements will be written in simple sorted order, which is faster, but may result in less readable output in Turtle or TriG.

.TP
.BR \-s " " \fIINPUT\fR
Parse \fIINPUT\fR as a string.

.TP
.BR \-t
Write terser output without newlines.

.TP
.BR \-v
Display version information and exit.

.SH AUTHOR
Serdi was written by David Robillard <d@drobilla.net>

.SH COPYRIGHT
Copyright \(co 2011-2020 David Robillard.
.br
License: <http://www.opensource.org/licenses/isc>
.br
This is free software; you are free to change and redistribute it.
.br
There is NO WARRANTY, to the extent permitted by law.

.SH "SEE ALSO"
<http://drobilla.net/software/serd>


NEWS
serd (1.0.1) unstable;

  * Add SerdBuffer for mutable buffers to keep SerdChunk const-correct
  * Make serd_strtod API const-correct
  * Remove serd_uri_to_path()
  * Remove useless character counting from API
  * Rename SerdChunk to SerdStringView
  * Use char* for strings in public API

 -- David Robillard <d@drobilla.net>  Sat, 19 Jan 2019 12:31:12 +0000

serd (0.30.7) unstable;

  * Deprecate serd_uri_to_path()
  * Don't install API man pages
  * Fix potential memory error when serialising URIs
  * Move headers to an include directory
  * Refuse to write relative URI references to NTriples
  * Remove use of C character class functions that may use locale
  * Split up and reorganize unit tests
  * Use aligned allocation via C11 or Windows API where possible

 -- David Robillard <d@drobilla.net>  Thu, 17 Dec 2020 09:09:31 +0000

serd (0.30.6) stable;

  * Fix various minor warnings and other code quality issues
  * Improve reader error handling
  * Make all free functions tolerant of NULL
  * Parse prefixed names more strictly

 -- David Robillard <d@drobilla.net>  Sun, 27 Sep 2020 10:51:38 +0000

serd (0.30.4) stable;

  * Fix EOF handling while reading in bulk or from strings
  * Fix lax handling of string errors
  * Fix reading from a null-delimited socket

 -- David Robillard <d@drobilla.net>  Sun, 26 Apr 2020 16:04:05 +0000

serd (0.30.2) stable;

  * Fix GCC 4 build
  * Fix colliding blank nodes when parsing TriG
  * Fix missing parse error messages
  * Fix parsing "a" abbreviation without padding whitespace
  * Fix parsing TriG graphs with several squashed trailing dots
  * Fix resolving some URIs against base URIs with no trailing slash
  * Improve build system and CI integration
  * Improve documentation

 -- David Robillard <d@drobilla.net>  Sun, 20 Oct 2019 22:28:01 +0000

serd (0.30.0) stable;

  * Add NTriples test suite
  * Add basic error tolerant parsing for line-based formats
  * Add serd_free for freeing memory allocated by serd
  * Add serdi option to write ASCII output
  * Clarify errors returned by serd_env_expand()
  * Clean up testing code
  * Fix building with MSVC
  * Fix length of base64 encoded nodes
  * Fix parsing local names that end with escaped dots
  * Fix reported error when reading statements with only a blank node
  * Improve relative URI serialisation
  * Make serdi guess input syntax from extension if unspecified
  * Make serdi syntax options case-insensitive
  * Report errors for all failing test cases
  * Support strict parsing of prefixed names
  * Upgrade to waf 2.0.9 and autowaf a328983
  * Use more efficient stack allocation and growth policies

 -- David Robillard <d@drobilla.net>  Sun, 22 Jul 2018 18:34:09 +0000

serd (0.28.0) stable;

  * Add serd_node_from_substring()
  * Add support for reading from a user provided callback
  * Fix hangs when reading corrupt UTF-8
  * Fix parsing of hex escapes in file URIs (thanks Johannes Mueller)
  * Fix strict parsing of abolute URI schemes
  * Gracefully handle applications that write corrupt UTF-8

 -- David Robillard <d@drobilla.net>  Fri, 21 Jul 2017 06:28:47 +0000

serd (0.26.0) stable;

  * Add support for TriG and NQuads
  * Add support for braindead SPARQL PREFIX and BASE directives

 -- David Robillard <d@drobilla.net>  Tue, 10 Jan 2017 01:10:35 +0000

serd (0.24.0) stable;

  * Add serd_node_new_relative_uri()
  * Fix construction and comparison of URIs with UTF-8 characters
  * Fix documentation generation
  * Fix potential out of bounds read
  * Fix unaligned memory access, undefined behaviour which breaks on ARM
  * Report I/O errors with message and return appropriate status code
  * Report missing statement separator errors
  * Update serdi man page

 -- David Robillard <d@drobilla.net>  Tue, 20 Sep 2016 02:21:47 +0000

serd (0.22.0) stable;

  * Add serd_reader_set_strict() and -l (lax) option to serdi to tolerate
    parsing URIs with escaped characters
  * Add support for Turtle named inline nodes extension
  * Fix clash resolution when a blank node ID prefix is set
  * Fix reading statements ending with a blank then dot with no space
  * Fix serd_reader_read_file() for URIs with escaped characters (spaces)
  * Fix serializing fractional decimals that would round up
  * Fix warnings when building with ISO C++ compilers
  * Remove dependence on fmax() to avoid portability issues
  * Report errors for invalid IRI characters and missing terminators
  * Show serdi errors in standard format
  * Upgrade to waf 1.8.14

 -- David Robillard <d@drobilla.net>  Thu, 08 Oct 2015 19:34:18 +0000

serd (0.20.0) stable;

  * Add -lm to pkg-config libs
  * Correctly handle posix_memalign failure
  * Don't write xsd:decimal literals to Turtle bare if they would not be read
    back with the same type
  * Fix const-correctness violation for reader input string
  * Fix possible crash in serd_writer_end_anon() when writing invalid lists
  * Generate blank names like _:b1 and _:B2 not _:genid1 _:docid2
  * Support new RDF 1.1 Turtle
  * Update to waf 1.7.14

 -- David Robillard <d@drobilla.net>  Fri, 08 Aug 2014 22:02:54 +0000

serd (0.18.2) stable;

  * Disable timestamps in HTML documentation for reproducible build
  * Fix bug that caused "a" abbreviation in non-predicate position
  * Fix clashing symbol "error" in amalgamation build
  * Fix crash when resolving against non-standard base URIs
  * Fix crash when serd_node_new_decimal is called with infinity or NaN
  * Update to waf 1.7.8 and autowaf r90 (install docs to versioned directory)

 -- David Robillard <d@drobilla.net>  Sun, 23 Dec 2012 02:32:15 +0000

serd (0.18.0) stable;

  * Add -e option to serdi to use incremental reading
  * Add -q option to serdi to suppress all non-data output, e.g. errors
  * Add error callback to reader and writer for custom error reporting
  * Add incremental read interface suitable for reading from infinite streams
  * Add option to build utilities as static binaries
  * Do not require a C++ compiler to build

src/serdi.c
/*
  Copyright 2011-2020 David Robillard <http://drobilla.net>

  Permission to use, copy, modify, and/or distribute this software for any
  purpose with or without fee is hereby granted, provided that the above
  copyright notice and this permission notice appear in all copies.

  THIS SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
*/

#define _POSIX_C_SOURCE 200809L /* for fileno and posix_fadvise */

#include "serd_config.h"
#include "string_utils.h"

#include "serd/serd.h"

#ifdef _WIN32
#define WIN32_LEAN_AND_MEAN 1
#include <fcntl.h>
#include <io.h>
#endif

#if defined(HAVE_POSIX_FADVISE) && defined(HAVE_FILENO)
#include <fcntl.h>
#endif

#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SERDI_ERROR(msg)       fprintf(stderr, "serdi: " msg)
#define SERDI_ERRORF(fmt, ...) fprintf(stderr, "serdi: " fmt, __VA_ARGS__)

typedef struct {
	SerdSyntax  syntax;
	const char* name;
	const char* extension;
} Syntax;

static const Syntax syntaxes[] = {
	{SERD_TURTLE,   "turtle",   ".ttl"},
	{SERD_NTRIPLES, "ntriples", ".nt"},
	{SERD_NQUADS,   "nquads",   ".nq"},
	{SERD_TRIG,     "trig",     ".trig"},
	{(SerdSyntax)0, NULL, NULL}
};

static SerdSyntax
get_syntax(const char* name)
{
	for (const Syntax* s = syntaxes; s->name; ++s) {
		if (!serd_strncasecmp(s->name, name, strlen(name))) {
			return s->syntax;
		}
	}
	SERDI_ERRORF("unknown syntax `%s'\n", name);
	return (SerdSyntax)0;
}

static SERD_PURE_FUNC SerdSyntax
guess_syntax(const char* filename)
{
	const char* ext = strrchr(filename, '.');
	if (ext) {
		for (const Syntax* s = syntaxes; s->name; ++s) {
			if (!serd_strncasecmp(s->extension, ext, strlen(ext))) {
				return s->syntax;
			}
		}
	}
	return (SerdSyntax)0;
}

static int
print_version(void)
{
	printf("serdi " SERD_VERSION " <http://drobilla.net/software/serd>\n");
	printf("Copyright 2011-2020 David Robillard <http://drobilla.net>.\n"
	       "License: <http://www.opensource.org/licenses/isc>\n"
	       "This is free software; you are free to change and redistribute it."
	       "\nThere is NO WARRANTY, to the extent permitted by law.\n");
	return 0;
}

static int
print_usage(const char* name, bool error)
{
	FILE* const os = error ? stderr : stdout;
	fprintf(os, "%s", error ? "\n" : "");
	fprintf(os, "Usage: %s [OPTION]... INPUT [BASE_URI]\n", name);
	fprintf(os, "Read and write RDF syntax.\n");
	fprintf(os, "Use - for INPUT to read from standard input.\n\n");
	fprintf(os, "  -a           Write ASCII output if possible.\n");
	fprintf(os, "  -b           Fast bulk output for large serialisations.\n");
	fprintf(os, "  -c PREFIX    Chop PREFIX from matching blank node IDs.\n");
	fprintf(os, "  -e           Eat input one character at a time.\n");
	fprintf(os, "  -f           Keep full URIs in input (don't qualify).\n");
	fprintf(os, "  -h           Display this help and exit.\n");
	fprintf(os, "  -i SYNTAX    Input syntax: turtle/ntriples/trig/nquads.\n");
	fprintf(os, "  -l           Lax (non-strict) parsing.\n");
	fprintf(os, "  -o SYNTAX    Output syntax: turtle/ntriples/nquads.\n");
	fprintf(os, "  -p PREFIX    Add PREFIX to blank node IDs.\n");
	fprintf(os, "  -q           Suppress all output except data.\n");
	fprintf(os, "  -r ROOT_URI  Keep relative URIs within ROOT_URI.\n");
	fprintf(os, "  -s INPUT     Parse INPUT as string (terminates options).\n");
	fprintf(os, "  -v           Display version information and exit.\n");
	return error ? 1 : 0;
}

static int
missing_arg(const char* name, char opt)
{
	SERDI_ERRORF("option requires an argument -- '%c'\n", opt);
	return print_usage(name, true);
}

static SerdStatus
quiet_error_sink(void* handle, const SerdError* e)
{
	(void)handle;
	(void)e;
	return SERD_SUCCESS;
}

static inline FILE*
serd_fopen(const char* path, const char* mode)
{
	FILE* fd = fopen(path, mode);
	if (!fd) {
		SERDI_ERRORF("failed to open file %s (%s)\n", path, strerror(errno));
		return NULL;
	}

#if defined(HAVE_POSIX_FADVISE) && defined(HAVE_FILENO)
	posix_fadvise(fileno(fd), 0, 0, POSIX_FADV_SEQUENTIAL|POSIX_FADV_NOREUSE);
#endif

	return fd;
}

static SerdStyle
choose_style(const SerdSyntax input_syntax,
             const SerdSyntax output_syntax,
             const bool       ascii,
             const bool       bulk_write,
             const bool       full_uris)
{
	unsigned output_style = 0u;
	if (output_syntax == SERD_NTRIPLES || ascii) {
		output_style |= SERD_STYLE_ASCII;
	} else if (output_syntax == SERD_TURTLE) {
		output_style |= SERD_STYLE_ABBREVIATED;
		if (!full_uris) {
			output_style |= SERD_STYLE_CURIED;
		}
	}

	if ((input_syntax == SERD_TURTLE || input_syntax == SERD_TRIG) ||
	    (output_style & SERD_STYLE_CURIED)) {
		// Base URI may change and/or we're abbreviating URIs, so must resolve
		output_style |= SERD_STYLE_RESOLVED;
	}

	if (bulk_write) {
		output_style |= SERD_STYLE_BULK;
	}

	return (SerdStyle)output_style;
}

int
main(int argc, char** argv)
{
	if (argc < 2) {
		return print_usage(argv[0], true);
	}

	FILE*          in_fd         = NULL;
	SerdSyntax     input_syntax  = (SerdSyntax)0;
	SerdSyntax     output_syntax = (SerdSyntax)0;
	bool           from_file     = true;
	bool           ascii         = false;
	bool           bulk_read     = true;
	bool           bulk_write    = false;
	bool           full_uris     = false;
	bool           lax           = false;
	bool           quiet         = false;
	const char*    in_name       = NULL;
	const char*    add_prefix    = NULL;
	const char*    chop_prefix   = NULL;
	const char*    root_uri      = NULL;
	int            a             = 1;
	for (; a < argc && argv[a][0] == '-'; ++a) {
		if (argv[a][1] == '\0') {
			in_name = (const char*)"(stdin)";
			in_fd   = stdin;
			break;
		} else if (argv[a][1] == 'a') {
			ascii = true;
		} else if (argv[a][1] == 'b') {
			bulk_write = true;
		} else if (argv[a][1] == 'e') {
			bulk_read = false;
		} else if (argv[a][1] == 'f') {
			full_uris = true;
		} else if (argv[a][1] == 'h') {
			return print_usage(argv[0], false);
		} else if (argv[a][1] == 'l') {
			lax = true;
		} else if (argv[a][1] == 'q') {
			quiet = true;
		} else if (argv[a][1] == 'v') {
			return print_version();
		} else if (argv[a][1] == 's') {
			in_name = (const char*)"(string)";
			from_file = false;
			++a;
			break;
		} else if (argv[a][1] == 'i') {
			if (++a == argc) {
				return missing_arg(argv[0], 'i');
			} else if (!(input_syntax = get_syntax(argv[a]))) {
				return print_usage(argv[0], true);
			}
		} else if (argv[a][1] == 'o') {
			if (++a == argc) {
				return missing_arg(argv[0], 'o');
			} else if (!(output_syntax = get_syntax(argv[a]))) {
				return print_usage(argv[0], true);
			}
		} else if (argv[a][1] == 'p') {
			if (++a == argc) {
				return missing_arg(argv[0], 'p');
			}
			add_prefix = argv[a];
		} else if (argv[a][1] == 'c') {
			if (++a == argc) {
				return missing_arg(argv[0], 'c');
			}
			chop_prefix = argv[a];
		} else if (argv[a][1] == 'r') {
			if (++a == argc) {
				return missing_arg(argv[0], 'r');
			}
			root_uri = argv[a];
		} else {
			SERDI_ERRORF("invalid option -- '%s'\n", argv[a] + 1);
			return print_usage(argv[0], true);
		}
	}

	if (a == argc) {
		SERDI_ERROR("missing input\n");
		return 1;
	}

#ifdef _WIN32
	_setmode(_fileno(stdin), _O_BINARY);
	_setmode(_fileno(stdout), _O_BINARY);
#endif

	char*       input_path = NULL;
	const char* input      = (const char*)argv[a++];
	if (from_file) {
		in_name = in_name ? in_name : input;
		if (!in_fd) {
			if (!strncmp(input, "file:", 5)) {
				input_path = serd_file_uri_parse(input, NULL);
				input      = input_path;
			}
			if (!input || !(in_fd = serd_fopen(input, "rb"))) {
				return 1;
			}
		}
	}

	if (!input_syntax && !(input_syntax = guess_syntax(in_name))) {
		input_syntax = SERD_TRIG;
	}

	if (!output_syntax) {
		output_syntax = (
			(input_syntax == SERD_TURTLE || input_syntax == SERD_NTRIPLES)
			? SERD_NTRIPLES
			: SERD_NQUADS);
	}

	const SerdStyle output_style =
	    choose_style(input_syntax, output_syntax, ascii, bulk_write, full_uris);

	SerdURI  base_uri = SERD_URI_NULL;
	SerdNode base     = SERD_NODE_NULL;
	if (a < argc) {  // Base URI given on command line
		base = serd_new_uri_from_string(
			(const char*)argv[a], NULL, &base_uri);
	} else if (from_file && in_fd != stdin) {  // Use input file URI
		base = serd_new_file_uri(input, NULL, &base_uri);
	}

	FILE* const    out_fd = stdout;
	SerdEnv* const env    = serd_env_new(&base);

	SerdWriter* const writer = serd_writer_new(
	    output_syntax, output_style, env, &base_uri, serd_file_sink, out_fd);

	SerdReader* const reader = serd_reader_new(
		input_syntax, writer, NULL,
		(SerdBaseSink)serd_writer_set_base_uri,
		(SerdPrefixSink)serd_writer_set_prefix,
		(SerdStatementSink)serd_writer_write_statement,
		(SerdEndSink)serd_writer_end_anon);

	serd_reader_set_strict(reader, !lax);
	if (quiet) {
		serd_reader_set_error_sink(reader, quiet_error_sink, NULL);
		serd_writer_set_error_sink(writer, quiet_error_sink, NULL);
	}

	SerdNode root = serd_node_from_string(SERD_URI, root_uri);
	serd_writer_set_root_uri(writer, &root);
	serd_writer_chop_blank_prefix(writer, chop_prefix);
	serd_reader_add_blank_prefix(reader, add_prefix);

	SerdStatus st = SERD_SUCCESS;
	if (!from_file) {
		st = serd_reader_read_string(reader, input);
	} else if (bulk_read) {
		st = serd_reader_read_file_handle(reader, in_fd, in_name);
	} else {
		st = serd_reader_start_stream(reader, in_fd, in_name, false);
		while (!st) {
			st = serd_reader_read_chunk(reader);
		}
		serd_reader_end_stream(reader);
	}

	serd_reader_free(reader);
	serd_writer_finish(writer);
	serd_writer_free(writer);
	serd_env_free(env);
	serd_node_free(&base);
	free(input_path);

	if (from_file) {
		fclose(in_fd);
	}

	if (fclose(out_fd)) {
		perror("serdi: write error");
		st = SERD_ERR_UNKNOWN;
	}

	return (st > SERD_FAILURE) ? 1 : 0;
}
